// react-native-web-lite/dist/esm/DeviceInfo/index.js
import { canUseDOM } from "react-native-web-internals";
import Dimensions from "../Dimensions/index";
const DeviceInfo = {
  Dimensions: {
    get windowPhysicalPixels() {
      const { width, height, fontScale, scale } = Dimensions.get("window");
      return {
        width: width * scale,
        height: height * scale,
        scale,
        fontScale
      };
    },
    get screenPhysicalPixels() {
      const { width, height, fontScale, scale } = Dimensions.get("screen");
      return {
        width: width * scale,
        height: height * scale,
        scale,
        fontScale
      };
    }
  },
  get locale() {
    if (canUseDOM) {
      if (navigator.languages) {
        return navigator.languages[0];
      } else {
        return navigator.language;
      }
    }
  },
  get totalMemory() {
    return canUseDOM ? navigator.deviceMemory : void 0;
  },
  get userAgent() {
    return canUseDOM ? navigator.userAgent : "";
  }
};
var DeviceInfo_default = DeviceInfo;
export {
  DeviceInfo_default as default
};
