// @tamagui/scroll-view/dist/cjs/ScrollView.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ScrollView_exports = {};
__export(ScrollView_exports, {
  ScrollView: () => ScrollView
});
module.exports = __toCommonJS(ScrollView_exports);
var import_stacks = require("@tamagui/stacks");
var import_web = require("@tamagui/web");
var import_react_native = require("react-native");
(0, import_web.setupReactNative)({
  ScrollView: import_react_native.ScrollView
});
const ScrollView = (0, import_web.styled)(import_react_native.ScrollView, {
  name: "ScrollView",
  scrollEnabled: true,
  variants: {
    fullscreen: {
      true: import_stacks.fullscreenStyle
    }
  }
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ScrollView
});
