// react-native-web-lite/dist/esm/AppRegistry/AppContainer.js
import { jsx } from "react/jsx-runtime";
import * as React from "react";
import { StyleSheet } from "react-native-web-internals";
import View from "../View/index";
const RootTagContext = React.createContext(null);
const AppContainer = React.forwardRef(
  (props, forwardedRef) => {
    const { children, WrapperComponent } = props;
    let innerView = /* @__PURE__ */ jsx(View, { pointerEvents: "box-none", style: styles.appContainer, children }, 1);
    if (WrapperComponent) {
      innerView = /* @__PURE__ */ jsx(WrapperComponent, { children: innerView });
    }
    return /* @__PURE__ */ jsx(RootTagContext.Provider, { value: props.rootTag, children: /* @__PURE__ */ jsx(
      View,
      {
        pointerEvents: "box-none",
        ref: forwardedRef,
        style: styles.appContainer,
        children: innerView
      }
    ) });
  }
);
AppContainer.displayName = "AppContainer";
var AppContainer_default = AppContainer;
const styles = StyleSheet.create({
  appContainer: {
    flex: 1
  }
});
export {
  AppContainer_default as default
};
