// react-native-web-lite/dist/esm/AppRegistry/index.js
import { unmountComponentAtNode } from "react-dom";
import { invariant } from "react-native-web-internals";
import renderApplication, { getApplication } from "./renderApplication";
const emptyObject = {};
const runnables = {};
let componentProviderInstrumentationHook = (component) => component();
let wrapperComponentProvider;
class AppRegistry {
  static getAppKeys() {
    return Object.keys(runnables);
  }
  static getApplication(appKey, appParameters) {
    var _a, _b;
    invariant(
      runnables[appKey] && runnables[appKey].getApplication,
      `Application ${appKey} has not been registered. This is either due to an import error during initialization or failure to call AppRegistry.registerComponent.`
    );
    return (_b = (_a = runnables[appKey]) == null ? void 0 : _a.getApplication) == null ? void 0 : _b.call(_a, appParameters);
  }
  static registerComponent(appKey, componentProvider) {
    runnables[appKey] = {
      getApplication: (appParameters) => getApplication(
        componentProviderInstrumentationHook(componentProvider),
        appParameters ? appParameters.initialProps : emptyObject,
        wrapperComponentProvider && wrapperComponentProvider(appParameters)
      ),
      run: (appParameters) => renderApplication(
        componentProviderInstrumentationHook(componentProvider),
        wrapperComponentProvider && wrapperComponentProvider(appParameters),
        appParameters.callback,
        {
          hydrate: appParameters.hydrate || false,
          initialProps: appParameters.initialProps || emptyObject,
          mode: appParameters.mode || "legacy",
          rootTag: appParameters.rootTag
        }
      )
    };
    return appKey;
  }
  static registerConfig(config) {
    config.forEach(({ appKey, component, run }) => {
      if (run) {
        AppRegistry.registerRunnable(appKey, run);
      } else {
        invariant(component, "No component provider passed in");
        AppRegistry.registerComponent(appKey, component);
      }
    });
  }
  // TODO: fix style sheet creation when using this method
  static registerRunnable(appKey, run) {
    runnables[appKey] = { run };
    return appKey;
  }
  static runApplication(appKey, appParameters) {
    const isDevelopment = process.env.NODE_ENV !== "production" && process.env.NODE_ENV !== "test";
    if (isDevelopment) {
      const params = { ...appParameters };
      params.rootTag = `#${params.rootTag.id}`;
      console.log(
        `Running application "${appKey}" with appParams:
`,
        params,
        `
Development-level warnings: ${isDevelopment ? "ON" : "OFF"}.
Performance optimizations: ${isDevelopment ? "OFF" : "ON"}.`
      );
    }
    invariant(
      runnables[appKey] && runnables[appKey].run,
      `Application "${appKey}" has not been registered. This is either due to an import error during initialization or failure to call AppRegistry.registerComponent.`
    );
    return runnables[appKey].run(appParameters);
  }
  static setComponentProviderInstrumentationHook(hook) {
    componentProviderInstrumentationHook = hook;
  }
  static setWrapperComponentProvider(provider) {
    wrapperComponentProvider = provider;
  }
  static unmountApplicationComponentAtRootTag(rootTag) {
    unmountComponentAtNode(rootTag);
  }
}
export {
  AppRegistry as default
};
