// react-native-web-lite/dist/esm/AppRegistry/renderApplication.js
import { jsx } from "react/jsx-runtime";
import { StyleSheet } from "react-native-web-internals";
import { invariant } from "react-native-web-internals";
import renderLegacy, { hydrate, hydrateLegacy, render } from "../render/index";
import AppContainer from "./AppContainer";
function renderApplication(RootComponent, WrapperComponent = null, callback = () => {
}, options) {
  const { hydrate: shouldHydrate, initialProps, mode, rootTag } = options;
  const renderFn = shouldHydrate ? mode === "concurrent" ? hydrate : hydrateLegacy : mode === "concurrent" ? render : renderLegacy;
  invariant(rootTag, "Expect to have a valid rootTag, instead got ", rootTag);
  return renderFn(
    /* @__PURE__ */ jsx(
      AppContainer,
      {
        WrapperComponent,
        ref: callback,
        rootTag,
        children: /* @__PURE__ */ jsx(RootComponent, { ...initialProps })
      }
    ),
    rootTag
  );
}
function getApplication(RootComponent, initialProps, WrapperComponent) {
  const element = /* @__PURE__ */ jsx(AppContainer, { WrapperComponent, rootTag: {}, children: /* @__PURE__ */ jsx(RootComponent, { ...initialProps }) });
  const getStyleElement = (props) => {
    const sheet = StyleSheet.getSheet();
    return /* @__PURE__ */ jsx(
      "style",
      {
        ...props,
        dangerouslySetInnerHTML: { __html: sheet.textContent },
        id: sheet.id
      }
    );
  };
  return { element, getStyleElement };
}
export {
  renderApplication as default,
  getApplication
};
