// tamagui/dist/cjs/views/TextArea.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TextArea_exports = {};
__export(TextArea_exports, {
  TextArea: () => TextArea,
  TextAreaFrame: () => TextAreaFrame
});
module.exports = __toCommonJS(TextArea_exports);
var import_core = require("@tamagui/core");
var import_focusable = require("@tamagui/focusable");
var import_inputHelpers = require("../helpers/inputHelpers");
var import_Input = require("./Input");
const TextAreaFrame = (0, import_core.styled)(import_Input.InputFrame, {
  name: "TextArea",
  multiline: true,
  variants: {
    unstyled: {
      false: {
        ...import_Input.defaultStyles,
        height: "auto",
        numberOfLines: 4
      }
    },
    size: {
      "...size": import_inputHelpers.textAreaSizeVariant
    }
  },
  defaultVariants: {
    unstyled: false
  }
});
const TextArea = (0, import_focusable.focusableInputHOC)(TextAreaFrame);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  TextArea,
  TextAreaFrame
});
