// react-native-web-lite/dist/cjs/DeviceInfo/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DeviceInfo_exports = {};
__export(DeviceInfo_exports, {
  default: () => DeviceInfo_default
});
module.exports = __toCommonJS(DeviceInfo_exports);
var import_react_native_web_internals = require("react-native-web-internals");
var import_Dimensions = __toESM(require("../Dimensions/index"));
const DeviceInfo = {
  Dimensions: {
    get windowPhysicalPixels() {
      const { width, height, fontScale, scale } = import_Dimensions.default.get("window");
      return {
        width: width * scale,
        height: height * scale,
        scale,
        fontScale
      };
    },
    get screenPhysicalPixels() {
      const { width, height, fontScale, scale } = import_Dimensions.default.get("screen");
      return {
        width: width * scale,
        height: height * scale,
        scale,
        fontScale
      };
    }
  },
  get locale() {
    if (import_react_native_web_internals.canUseDOM) {
      if (navigator.languages) {
        return navigator.languages[0];
      } else {
        return navigator.language;
      }
    }
  },
  get totalMemory() {
    return import_react_native_web_internals.canUseDOM ? navigator.deviceMemory : void 0;
  },
  get userAgent() {
    return import_react_native_web_internals.canUseDOM ? navigator.userAgent : "";
  }
};
var DeviceInfo_default = DeviceInfo;
