// @tamagui/form/dist/jsx/Form.mjs
import {
  Stack,
  composeEventHandlers,
  styled,
  withStaticProperties
} from "@tamagui/core";
import { createContextScope } from "@tamagui/create-context";
import { cloneElement } from "react";
const FORM_NAME = "Form";
const FormFrame = styled(Stack, {
  name: FORM_NAME,
  tag: "form"
});
const [createFormContext] = createContextScope(FORM_NAME);
const [FormProvider, useFormContext] = createFormContext(FORM_NAME);
const TRIGGER_NAME = "FormTrigger";
const FormTriggerFrame = styled(Stack, {
  name: TRIGGER_NAME
});
const FormTrigger = FormTriggerFrame.styleable(
  (props, forwardedRef) => {
    const { __scopeForm, children, ...triggerProps } = props;
    const context = useFormContext(TRIGGER_NAME, __scopeForm);
    return <FormTriggerFrame
      tag="button"
      {...triggerProps}
      children={triggerProps.asChild ? cloneElement(children, { disabled: triggerProps.disabled }) : children}
      ref={forwardedRef}
      onPress={composeEventHandlers(props.onPress, context.onSubmit)}
    />;
  }
);
function FormComponent({
  onSubmit,
  ...props
}) {
  return <FormProvider scope={props.__scopeForm} onSubmit={onSubmit}><FormFrame {...props} onSubmit={(e) => e.preventDefault()} /></FormProvider>;
}
const Form = withStaticProperties(FormFrame.extractable(FormComponent), {
  Trigger: FormTrigger
});
export {
  Form,
  FormFrame,
  FormProvider,
  FormTrigger,
  useFormContext
};
