// react-native-web-lite/dist/cjs/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __reExport = (target, mod, secondTarget) => (__copyProps(target, mod, "default"), secondTarget && __copyProps(secondTarget, mod, "default"));
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var src_exports = {};
__export(src_exports, {
  AccessibilityInfo: () => import_AccessibilityInfo.default,
  ActivityIndicator: () => import_ActivityIndicator.default,
  Alert: () => import_Alert.default,
  Animated: () => import_Animated.default,
  AppRegistry: () => import_AppRegistry.default,
  AppState: () => import_AppState.default,
  Appearance: () => import_Appearance.default,
  BackHandler: () => import_BackHandler.default,
  Clipboard: () => import_Clipboard.default,
  DeviceEmitter: () => import_DeviceEmitter.default,
  DeviceEventEmitter: () => import_DeviceEmitter2.default,
  DeviceInfo: () => import_DeviceInfo.default,
  Dimensions: () => import_Dimensions.default,
  DrawerLayoutAndroid: () => import_UnimplementedView.default,
  Easing: () => import_Easing.default,
  FlatList: () => import_UnimplementedView4.default,
  I18nManager: () => import_I18nManager.default,
  Image: () => import_Image.default,
  ImageBackground: () => import_ImageBackground.default,
  ImageLoader: () => import_react_native_web_internals.ImageLoader,
  Keyboard: () => import_Keyboard.default,
  KeyboardAvoidingView: () => import_KeyboardAvoidingView.default,
  Linking: () => import_Linking.default,
  LogBox: () => import_LogBox.default,
  Modal: () => import_Modal.default,
  NativeEventEmitter: () => import_NativeEventEmitter.default,
  NativeModules: () => import_NativeModules.default,
  PanResponder: () => import_PanResponder.default,
  PixelRatio: () => import_PixelRatio.default,
  Platform: () => import_react_native_web_internals.Platform,
  Pressable: () => import_Pressable.default,
  RefreshControl: () => import_RefreshControl.default,
  SafeAreaView: () => import_SafeAreaView.default,
  ScrollView: () => import_ScrollView.default,
  SectionList: () => import_UnimplementedView7.default,
  Share: () => import_Share.default,
  StatusBar: () => import_StatusBar.default,
  Switch: () => import_UnimplementedView2.default,
  Text: () => import_Text.default,
  TextInput: () => import_TextInput.default,
  TouchableHighlight: () => import_UnimplementedView5.default,
  TouchableNativeFeedback: () => import_UnimplementedView6.default,
  TouchableOpacity: () => import_TouchableOpacity.default,
  TouchableWithoutFeedback: () => import_TouchableWithoutFeedback.default,
  UIManager: () => import_react_native_web_internals.UIManager,
  Vibration: () => import_Vibration.default,
  View: () => import_View.default,
  VirtualizedList: () => import_UnimplementedView3.default,
  canUseDOM: () => import_react_native_web_internals.canUseDOM,
  dismissKeyboard: () => import_react_native_web_internals.dismissKeyboard,
  findNodeHandle: () => import_findNodeHandle.findNodeHandle,
  isWebColor: () => import_react_native_web_internals.isWebColor,
  mergeRefs: () => import_react_native_web_internals.mergeRefs,
  processColor: () => import_react_native_web_internals.processColor,
  render: () => import_render.default,
  requireNativeComponent: () => requireNativeComponent,
  unmountComponentAtNode: () => import_react_dom.unmountComponentAtNode,
  unstable_createElement: () => import_createElement.default,
  useColorScheme: () => import_useColorScheme.default,
  useEvent: () => import_react_native_web_internals.useEvent,
  useHover: () => import_react_native_web_internals.useHover,
  useLocaleContext: () => import_useLocaleContext.useLocaleContext,
  useMergeRefs: () => import_react_native_web_internals.useMergeRefs,
  useWindowDimensions: () => import_useWindowDimensions.default
});
module.exports = __toCommonJS(src_exports);
var import_createElement = __toESM(require("./createElement/index"));
var import_react_native_web_internals = require("react-native-web-internals");
var import_render = __toESM(require("./render/index"));
var import_react_dom = require("react-dom");
var import_NativeModules = __toESM(require("./NativeModules/index"));
var import_findNodeHandle = require("./findNodeHandle");
var import_Easing = __toESM(require("./vendor/react-native/Animated/Easing"));
var import_Animated = __toESM(require("./vendor/react-native/Animated/Animated"));
var import_NativeEventEmitter = __toESM(require("./vendor/react-native/NativeEventEmitter/index"));
var import_AccessibilityInfo = __toESM(require("./AccessibilityInfo/index"));
var import_Alert = __toESM(require("./Alert/index"));
var import_Appearance = __toESM(require("./Appearance/index"));
var import_AppRegistry = __toESM(require("./AppRegistry/index"));
var import_AppState = __toESM(require("./AppState/index"));
var import_BackHandler = __toESM(require("./BackHandler/index"));
var import_Clipboard = __toESM(require("./Clipboard/index"));
var import_DeviceInfo = __toESM(require("./DeviceInfo/index"));
var import_DeviceEmitter = __toESM(require("./DeviceEmitter"));
var import_DeviceEmitter2 = __toESM(require("./DeviceEmitter"));
var import_Dimensions = __toESM(require("./Dimensions/index"));
var import_I18nManager = __toESM(require("./I18nManager/index"));
var import_Keyboard = __toESM(require("./Keyboard/index"));
var import_Linking = __toESM(require("./Linking/index"));
var import_PanResponder = __toESM(require("./PanResponder/index"));
var import_PixelRatio = __toESM(require("./PixelRatio/index"));
var import_Share = __toESM(require("./Share/index"));
var import_Vibration = __toESM(require("./Vibration/index"));
var import_UnimplementedView = __toESM(require("./UnimplementedView"));
var import_UnimplementedView2 = __toESM(require("./UnimplementedView"));
var import_UnimplementedView3 = __toESM(require("./UnimplementedView"));
var import_UnimplementedView4 = __toESM(require("./UnimplementedView"));
var import_UnimplementedView5 = __toESM(require("./UnimplementedView"));
var import_UnimplementedView6 = __toESM(require("./UnimplementedView"));
var import_UnimplementedView7 = __toESM(require("./UnimplementedView"));
var import_TouchableOpacity = __toESM(require("./TouchableOpacity"));
var import_TouchableWithoutFeedback = __toESM(require("./TouchableWithoutFeedback"));
var import_ActivityIndicator = __toESM(require("./ActivityIndicator/index"));
var import_Image = __toESM(require("./Image/index"));
var import_ImageBackground = __toESM(require("./ImageBackground/index"));
var import_KeyboardAvoidingView = __toESM(require("./KeyboardAvoidingView/index"));
var import_Modal = __toESM(require("./Modal/index"));
var import_Pressable = __toESM(require("./Pressable/index"));
var import_RefreshControl = __toESM(require("./RefreshControl/index"));
var import_SafeAreaView = __toESM(require("./SafeAreaView/index"));
var import_ScrollView = __toESM(require("./ScrollView/index"));
var import_StatusBar = __toESM(require("./StatusBar/index"));
var import_Text = __toESM(require("./Text/index"));
var import_TextInput = __toESM(require("./TextInput/index"));
var import_View = __toESM(require("./View/index"));
var import_LogBox = __toESM(require("./LogBox/index"));
var import_useColorScheme = __toESM(require("./useColorScheme/index"));
var import_useLocaleContext = require("./useLocaleContext/index");
var import_useWindowDimensions = __toESM(require("./useWindowDimensions/index"));
__reExport(src_exports, require("react-native-web-internals"), module.exports);
function requireNativeComponent(name) {
  return function FakeComponent() {
    return null;
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AccessibilityInfo,
  ActivityIndicator,
  Alert,
  Animated,
  AppRegistry,
  AppState,
  Appearance,
  BackHandler,
  Clipboard,
  DeviceEmitter,
  DeviceEventEmitter,
  DeviceInfo,
  Dimensions,
  DrawerLayoutAndroid,
  Easing,
  FlatList,
  I18nManager,
  Image,
  ImageBackground,
  ImageLoader,
  Keyboard,
  KeyboardAvoidingView,
  Linking,
  LogBox,
  Modal,
  NativeEventEmitter,
  NativeModules,
  PanResponder,
  PixelRatio,
  Platform,
  Pressable,
  RefreshControl,
  SafeAreaView,
  ScrollView,
  SectionList,
  Share,
  StatusBar,
  Switch,
  Text,
  TextInput,
  TouchableHighlight,
  TouchableNativeFeedback,
  TouchableOpacity,
  TouchableWithoutFeedback,
  UIManager,
  Vibration,
  View,
  VirtualizedList,
  canUseDOM,
  dismissKeyboard,
  findNodeHandle,
  isWebColor,
  mergeRefs,
  processColor,
  render,
  requireNativeComponent,
  unmountComponentAtNode,
  unstable_createElement,
  useColorScheme,
  useEvent,
  useHover,
  useLocaleContext,
  useMergeRefs,
  useWindowDimensions,
  ...require("react-native-web-internals")
});
